src/net/primitives.c

```diff
@@ -314,12 +314,13 @@ int8_t _z_send_reply(const _z_query_t *query, _z_keyexpr_t keyexpr, const _z_val
                      const z_sample_kind_t kind, const z_congestion_control_t cong_ctrl, z_priority_t priority,
                      _Bool is_express, const _z_timestamp_t *timestamp, const _z_bytes_t att) {
     int8_t ret = _Z_RES_OK;
+    _z_session_t *zn = &query->_zn.in->val;
```

Comment:
should this one also use upgrade instead of getting the pointer directly

Reply:
I think the pointer should be upgraded in the api layer, before this call so we can bail quickly if it fails.

Reply:
then strong rc should be passed explicitly, otherwise when reading the code it is unclear what guarantees that reading ptr from weak is ok.